### PR TITLE
refactor(experimental): add codecs-numbers package

### PR DESCRIPTION
**Pull request by lorisleiva**
This PR adds number codecs to web3.js. It adds the following functions `getXEncoder`, `getXDecoder` and `getXCodec` for each of the following `X`s:

- `F32`
- `F64`
- `I128`
- `I16`
- `I32`
- `I64`
- `I8`
- `ShortU16`
- `U128`
- `U16`
- `U32`
- `U64`
- `U8`

### CODE REVIEW

**Comment by lorisleiva** [comment]
Current dependencies on/for this PR:
* master
  * **PR #1630** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/1630" target="_blank">Graphite</a> 
    * **PR #1631** <a href="https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/1631" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/1631?utm_source=stack-comment).

File: packages/codecs-numbers/src/assertions.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+ */
+export function assertNumberIsBetweenForCodec(
+    codecDescription: string,
+    min: number | bigint,
+    max: number | bigint,
+    value: number | bigint
+) {
+    if (value < min || value > max) {
+        // TODO: Coded error.
+        throw new Error(`Codec [${codecDescription}] expected number to be between ${min} and ${max}, got ${value}.`);
```
Comment (steveluscher) [inline packages/codecs-numbers/src/assertions.ts:12]: Maybe ‘between … inclusive’ or use math-y syntax ‘in the range [min, max]’

File: packages/codecs-numbers/src/common.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+     * @defaultValue `Endian.Little`
+     */
+    endian?: Endian;
+};
+
+/** Defines the endianness of a number serializer. */
+export enum Endian {
+    Little = 'le',
+    Big = 'be',
+}
```
Comment (steveluscher) [inline packages/codecs-numbers/src/common.ts:28]: Can we make this a `const enum`? No reason to cargo cult the strings around post-compile, right?

**Review by steveluscher — APPROVED** [review]

File: packages/codecs-numbers/src/assertions.ts (hunk 1 above)
Comment (lorisleiva) [reply packages/codecs-numbers/src/assertions.ts]: Done ✅

File: packages/codecs-numbers/src/common.ts (hunk 2 above)
Comment (lorisleiva) [reply packages/codecs-numbers/src/common.ts]: Done ✅

**Comment by lorisleiva** [comment]
### Merge Activity

* **Sep 30, 6:16 PM**: @lorisleiva started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/solana-labs/solana-web3.js/1631).

**Pull request merged**